{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import operator as op\n",
    "from collections import Counter\n",
    "from itertools import chain, groupby, takewhile\n",
    "from typing import Optional, Callable, Tuple, List\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "\n",
    "import regex as re\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import tqdm\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_rna, generic_dna\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from fn import F\n",
    "\n",
    "sns.set(color_codes=True)\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Region extraction utilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALPHABET = {\n",
    "    'A': 'A',\n",
    "    'C': 'C',\n",
    "    'G': 'G',\n",
    "    'T': 'T',\n",
    "    'R': '[AG]',\n",
    "    'Y': '[CT]',\n",
    "    'S': '[GC]',\n",
    "    'W': '[AT]',\n",
    "    'K': '[GT]',\n",
    "    'M': '[AC]',\n",
    "    'B': '[CGT]',\n",
    "    'D': '[AGT]',\n",
    "    'H': '[ACT]',\n",
    "    'V': '[ACG]',\n",
    "    'N': '[ACGT]'\n",
    "}\n",
    "\n",
    "Primer = Callable[[str], Optional[Tuple[int, int]]]\n",
    "\n",
    "\n",
    "def mkprimer(substitutions: int, primer_sequence: str) -> Primer:\n",
    "    \"\"\"\n",
    "    Create a universal primer.\n",
    "    :param substitutions: the number of allowed mismatches\n",
    "    :param primer: An IUPAC-encoded primer sequence\n",
    "    :return: A callable that searches for the first occurrence of the primer\n",
    "             in a sequence and returns it's span (if found).\n",
    "    :raises ValueError: primer sequence contains non-IUPAC DNA characters.  \n",
    "    \"\"\"\n",
    "    try:\n",
    "        base = ''.join(ALPHABET[base] for base in primer_sequence)\n",
    "        fuzzy = f'{{s<={substitutions}}}' if substitutions else ''\n",
    "        pattern = re.compile(f'(:?{base}){fuzzy}', flags=re.BESTMATCH)\n",
    "    except KeyError as err:\n",
    "        raise ValueError(f'unknown base: {err}')\n",
    "    \n",
    "    def primer(sequence: str) -> Optional[Tuple[int, int]]:\n",
    "        match = pattern.search(sequence)\n",
    "        return match.span() if match else None\n",
    "    \n",
    "    return primer\n",
    "\n",
    "\n",
    "def extract_region(forward: Primer, reverse: Primer, seqrec: SeqRecord) -> Optional[SeqRecord]:\n",
    "    \"\"\"\n",
    "    Extract a region flanked by the primers \n",
    "    \"\"\"\n",
    "    try:\n",
    "        _, start = forward(str(seqrec.seq))\n",
    "        end, _ = reverse(str(seqrec.seq))\n",
    "    except TypeError: # can't find primers\n",
    "        return None\n",
    "    return seqrec[start:end]\n",
    "\n",
    "\n",
    "# Standard 16S rRNA V4-region forward and reverse universal primers; allow up to 2 mismatches\n",
    "fwd_primer = mkprimer(2, 'GTGCCAGCMGCCGCGGTAA')\n",
    "rev_primer = mkprimer(2, str(Seq('GGACTACVSGGGTATCTAAT', generic_dna).reverse_complement()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read sequences, discard short records, extract V4 regions, discard failed records, parse taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def back_transcribe(seqrec: SeqRecord) -> SeqRecord:\n",
    "    \"\"\"\n",
    "    Back-transcribe an RNA sequence record\n",
    "    \"\"\"\n",
    "    return SeqRecord(id=seqrec.id, name=seqrec.name, description=seqrec.description,\n",
    "                     seq=seqrec.seq.back_transcribe())\n",
    "\n",
    "def parse_taxonomy(seqrec: SeqRecord) -> List[str]:\n",
    "    \"\"\"\n",
    "    Parse taxonomy from SILVA sequence records\n",
    "    \"\"\"\n",
    "    return seqrec.description.split(' ', 1)[1].split(';')\n",
    "\n",
    "\n",
    "with gzip.open('taxonomy/SILVA_132_SSURef_tax_silva_trunc.fasta.gz', 'rt') as silva_buffer:\n",
    "    with ThreadPoolExecutor(4) as workers:\n",
    "        silva_seqs = (\n",
    "            F(filter, lambda seqrec: 1100 <= len(seqrec) <= 1600) >>\n",
    "            (map, back_transcribe) >> \n",
    "            (workers.map, F(extract_region, fwd_primer, rev_primer)) >> \n",
    "            (filter, bool) >> \n",
    "            list\n",
    "        )(SeqIO.parse(silva_buffer, 'fasta'))\n",
    "        silva_taxa = list(map(parse_taxonomy, silva_seqs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot sequence length distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda/envs/students/lib/python3.6/site-packages/scipy/stats/stats.py:1713: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.\n",
      "  return np.add.reduce(sorted[indexer] * weights, axis=axis) / sumval\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(100, 400)]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXIAAAEBCAYAAABlki5mAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4xLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvDW2N/gAAFCNJREFUeJzt3VuMJNddx/FfVffMdM/OdPWsdxecizEiybGIwMTYIg5xEJElgoTFLSJEBhykBAgIyw+RkJAQBCmKFfIEJDIiQoKEREJCCjw5kR8i4mCLONghgehwkY2Nsdi1t2+zM90z3VU8VNfM7Mz0baZ6es6p70dazU51d/U56q7fnj11/lVBkiQCALgrXHQDAACnQ5ADgOMIcgBwHEEOAI4jyAHAcQQ5ADiOIAcAxxHkAOA4ghwAHEeQA4DjCHIAcFw5p/2sSLpH0iuSBjntEwB8V5J0q6SvS+qddCd5Bfk9kr6a074AoGjuk/TkSV+cV5C/IkmNxg3FsZ9XU7zlljW99trmopsxFz73TaJ/rvO5f2EYaGPjgjTM0JPKK8gHkhTHibdBLom+OYz+uc33/umUU9Kc7AQAxxHkAOA4ghwAHEeQA4DjCHIAcBxBDgCOI8gBwHEEOQA4jiAHAMcR5ADgOIIcABxHkAOA4whyAHAcQQ4AjiPIAcBxBDkAOI4gBwDHEeQonG63qyTx/o4zKBCCHIWyudnRI498WN/85rOLbgqQG4IchXLt2lXt7PT08ssvLbopQG4IchRKq9W86SfgA4IchdJut4c/WwtuCZAfghyFko3Em02CHP4gyFEo+0HeWHBLgPwQ5CiUVisdiTO1Ap8Q5CiUbETe7W5rZ2dnwa0B8kGQo1Aazf3VKozK4QuCHIXSabcULK9LYgki/EGQozC63a52dnoqrdQlEeTwB0GOwsiCO6xsDH9nagV+IMhRGFmQl1aim34HXEeQozCyk5vB0qpKSxVOdsIbBDkKIxuBB+WqVKowtQJvEOQojDS4AwWlZam0oiZTK/AEQY7CaLdbKi1VFAShgnJVzSZBDj8Q5CiMVqsplSqSpKBUUafd4k5B8AJBjsJoNveDPCxX1O/vant7e8GtAk6PIEdhNFtNBeXhiHz4kyWI8AFBjkKI41idTlvhoSBnCSJ8QJCjELa2bigeDA6MyKuSGJHDDwQ5CiFbM54FeFhiagX+IMhRCHvFQMMAV2lZQRAytQIvEOQohP2qzuHUShAoXKpS3QkvEOQohCyws5OdkqTSClMr8AJBjkJot5sKwpIULu1vLFWo7oQXCHIUQqvVUliuKgiCvW1hucL1VuAFghyFkJbnr9y0LShVtLnZURzHC2oVkA+CHIXQaDb3V6wMBeWKkjjW5mZnQa0C8kGQoxDardbeipXMflEQK1fgNoIc3uv3+7pxY/OYIKcoCH4gyOG9dMSd7I3AMyFBDk8Q5PBeo9GQpJEjcqo74TqCHN7L1oqHh4M8XFIQlpkjh/MIcnhvb0R+aNWKJIVLVbXbTK3AbQQ5vDdqakWSFK6o2WREDrcR5PBeo9FQWFpWEJaPPBaUK2o2GwtoFZAfghzeazQax4/Gla4lb3GyE44jyOG9RqNxpDw/E5RWtL11Q7u7u2fcKiA/BDm8d/1648ga8ky2vdNpn2WTgFwR5PBes9k4dsWKRFEQ/ECQw2u9Xk/dbnfMHDlBDvcR5PBaVrV5uBgosx/knPCEuwhyeO3wvToPy6ZcKNOHywhyeC0baY882RmWFJa5dyfcRpDDa1n5/aiTnVI67cLUClxGkMNr+yPy49eRS1ISMiKH2whyeK3Vaqq0VFUQjP6qB+WqGk2CHO4iyOG1Vqs1dlpFSk+EdjjZCYcR5PBas9lQMqI8PxOUK9rZSdebAy4iyOG1Zqs5culhJixRFAS3EeTwVpIk6rTbI4uBMlR3wnUEOby1tXVDg0FfQen4NeSZbI05RUFwFUEOb+0vPZx2RE6Qw00EObw1qTw/E5SWpSBgagXOIsjhrWyqZGKQB6FK5SpTK3AWQQ5vZVMl4YjrrNyE663AYQQ5vNVqNdOKznBp8pPDippUd8JRBDm81W63FC5VFQTBxOcG5YqajMjhKIIc3mq1miNvunxYWK6o02krjuM5twrIH0EObzWaTWnCdVYyQbmieDDQ1taNObcKyB9BDm+1Ws2JVZ2ZrCiIteRwEUEOL8VxrBubnZF3Bjos4HorcBhBDi91Om0lSTLxEraZkOutwGEEObw0bVVnJnseRUFwEUEOL81UDCRJ4ZKCsMQcOZxEkMNLM4/Ig0DhUpWpFTiJIIeXpr3y4U24CTMcRZDDS61WU2FpSUFYnvo1QYkyfbiJIIeX2u3W1EsPM0G5whw5nESQw0uzlOdngnJFN25sqt/vz6lVwHwQ5PBSs9mceg15Jh3BJ+p0OvNpFDAnBDm81Go1ZzvRqYNryZknh1sIcnhnd3dH3e72zEFOdSdcRZDDOydaeqiD11vhhCfcQpDDO3tVnaXZV61IlOnDPQQ5vJPNcc88Ig/LCkvLTK3AOQQ5vHPSqRUpnSdnagWuIcjhnVmvs3JQUqJMH+4hyOGd7M5AQTD71zsoVdRoNubQKmB+CHJ4p9VqTX2Lt8OCclVtplbgGIIc3mm1mkpmLM/PBOWKer2uer1ezq0C5ocgh3eardnL8zMhSxDhIIIcXkmS5ERXPswEVHfCQQQ5vNLtbqu/u3uqOXKJETncMv1V96fw3HPf0PZ2N89dnhu1WlXt9vZUzw2CYORjSZKcui3j9n8Ss/RtGtP08TR9GLf/LICPW3o42HpV/a2rKq9eUWn10vHtGk7JPPPMP2l3d3fk+8z7M57lPfL+/GZ9/3m9R2Ze/TsPqtWK3v3ud516P7kG+ec//1e6evVqnrsETiaJb/p1sPWq4mtP6T33368vP/GEdPneY8M8KC9Lkp5++mt6+umvnUlTUVxXrlw5f0F+551vO1VVXJJIOQ82c7OysqReb/QILTPNQOU0fcxpIHSTlZWyer18b6Ywro959GHU/judjr7znX+VDt3irb91Ve+5/3596EMflCQ9/tTzxwZ5Mkg/4yiqy5g7jn2PPD7jSd/1Wd7juO/mNPvP6/1PatrvwTy+n+dFFEW57CfXIH/wwQ8ojueQNOfA5cvrunbNzxsO+NS3zc1NPfzwrynp3/xf8fLqlXQkLunLTzyh8uV7j3190k+nBh988CHdffePzLexOfHp8zuOz/0Lw3xGrpzshFcuXLigsFTaC+RMafWSwsv36vGnnlc4YlpF2g/yWi2fkRJwFnIdkQOLFgSBauuRNvtHT7qXVi+NDPBMMkhH8nn9lxc4C4zI4Z0oqh8ZkU9rf0Rez7NJwFwR5PDOxkZdGpwsyON+V0tLy6pUTrYOHVgEghzeqdWiEwd50u+qVotyX6sPzBNBDu/UapEGu10lh9aSTyPpb6teZ1oFbiHI4Z0oqktKlAx2Zn5tEPcIcjiHIId30iDXkbXk04j7XU50wjkEObyTLR2cdeVKkgwU93ssPYRzCHJ4Z39EPmOQ93s3vR5wBUEO72RVmfHMQZ4+nxE5XEOQwzuVSkXLyyt7VZrTyubUGZHDNQQ5vFSLopmnVmKuswJHEeTwUv0EZfrJgCCHmwhyeKleryuIezO9Jul3Va2uamlpaU6tAuaDIIeXoqh+gpOd26oxPw4HEeTwUq1WV9zvKYkHU78m6Xe1QVUnHESQw0t7RUEzXDwriHusWIGTCHJ4KTthOcsJz7jfZQ05nESQw0uzVncm8a7iwS4jcjiJIIeXspF1POWFs7hXJ1xGkMNLs06txHvl+YzI4R6CHF4ql8taXb0w/dQKI3I4jCCHt6KoPvWqlYQRORxGkMNb9fr0ZfpJf1tBEGh9fX3OrQLyR5DDW1FUl+Ipg3zQ1drausKQQwLu4VsLb0VRXfFuV0mSTHxuuoacaRW4iSCHt6IoUhL3pbg/+ckDbroMdxHk8NZMN2EeMCKHuwhyeGvvlm8TVq4kSaJ4d5sgh7MIcnhr6jL9eEdJEnOdFTiLIIe3pq3upKoTriPI4a21tTUFYTgxyKnqhOsIcngrDEOtr9UmnuzMHmdqBa4iyOG1qD75lm+U58N1BDm8tlGvSxNuwhz3uyqVy6pWV8+oVUC+CHJ4rVarS5OWHw66qq1HCoLgjFoF5Isgh9eiKNJgd3tsmX7S36aqE04jyOG1KKpLSaJkMHp6JRjsEORwGkEOr2UrUcYtQUwGVHXCbQQ5vDapujNJYg12u6whh9MIcnhtb0Q+OH4tedLvDZ/HiBzuIsjhtVptwoh8kK0hZ0QOdxHk8FqlUtHy8vLIoqCsqpOpFbiMIIfXgiBIb8I8Msip6oT7CHJ47+LFjZHXW4m5YBY8QJDDexcvXlQwokw/6Xe1slLRysrKGbcKyA9BDu/Vx1w4K+lvq8aJTjiOIIf3NjY2FPd7SpLBkceSQVcb9Y0FtArID0EO721spEGdrRk/KBj0ONEJ5xHk8N5+kB+dXon7XdaQw3kEOby3H+Q3r1xJ4r7iwQ4jcjiPIIf3sisbHj7hyb064QuCHN7bG5EfusEE5fnwBUEO7y0tLalaXT0yRx5T1QlPEOQohFpUPzpHvje1QpDDbQQ5CmGjfvR6K2mwB1pfX19Mo4CcEOQohFotkg6V6Sf9ri5cWFO5XF5Qq4B8EOQohCiqK949OrXCiU74gCBHIURRpCTuK4l397Ylgy43XYYXCHIUwrH37qQ8H54gyFEI2RRKtuQwSRLF/W2CHF4gyFEIR+7dGe8qiQfMkcMLBDkK4fDUCuX58AlBjkJYX19XEAR7RUFUdcInBDkKIQxDra2t711fZf86KwQ53EeQozCiaP+Wb9nInDly+IAgR2HU63VpsD9HHpZKWl29sOBWAadHkKMwoqguDdIy/bjf1fp6TWHIIQD38S1GYWRl+kmSKOl3VWd+HJ4gyFEYURQpSWJpsCPFlOfDHwQ5CiNbMx4PulKf8nz4gyBHYewVBe1uadDfphgI3iDIURh711vptaUkYUQObxDkKIwsuAe9xvB3RuTwA0GOwqhWV1UqlxV3m5K4zgr8QZCjMIIgUG09UtxrSaI8H/4gyFEo6ZLDRBJBDn8Q5CiUbO348vKKKpXKglsD5IMgR6Fko/B15sfhEYIchZKd4NygqhMeIchRKNmInPlx+IQgR6Fka8dZQw6fEOQoFEbk8BFBjkK5ePGSgiDQpUuXF90UIDflRTcAOEsbGxv66Ec/rltvff2imwLkhiBH4bzhDbctuglArphaAQDHEeQA4DiCHAAcR5ADgOMIcgBwHEEOAI4jyAHAcQQ5ADiOIAcAxxHkAOA4ghwAHEeQA4DjCHIAcBxBDgCOI8gBwHEEOQA4jiAHAMcR5ADguLxu9VaSpDAMctrd+eRz/3zum0T/XOdr/w70q3Sa/QRJkpy+NdI7JX01jx0BQAHdJ+nJk744ryBfkXSPpFckDfLYIQAUQEnSrZK+Lql30p3kFeQAgAXhZCcAOI4gBwDHEeQA4DiCHAAcR5ADgOMIcgBwHEEOAI6bWKJvjPmkpJ+XdLukH7DWfnu4/S2S/lLSLZJek/Qr1tr/mPTYeTOmfy9I6g7/SNLvWGu/NHzs7ZL+TFJV0guSfslae/Us2z0tY8wtkj4r6fuUFhz8p6Rft9ZeG9cPV/o4oX+JpG9JiodP/2Vr7beGr3tA0h8pPQa+IelXrbVbZ93+SYwxX5T0vUr7sCnpt621z3l0/I3q3wvy4PjLGGN+X9IfaJgxeR9704zIvyjpXZL++9D2xyR9ylr7FkmfGr7xNI+dN6P6J0nvtdb+0PBP9iUKJH1O0m8N+/cPkh49s9bOLpH0CWutsdb+oKT/kvTouH441sdj+3fg8Xcc+AyzEF+T9OeSHrDWvklSR9JHzrrhU3rIWnuntfZtkj4p6S+G2305/kb1T/Lj+JMx5i5Jb5f04vD33I+9iUFurX3SWvvSoYZdkXSXpC8MN31B0l3GmMvjHpv0XotwXP8muFtS11qbXRfhMUm/kH/L8mGtvW6t/cqBTU9L+h6N74czfRzTv3F+UtIzB0apj0l63xyad2rW2taBXyNJsWfH35H+TXiJM99NSTLGrCj9x/Q3lQ46pDkceyedI3+jpJettQNJGv783+H2cY+55q+NMf9ijPm0MaY+3HabDozerbWvSgqNMRcX0sIZGGNCSR+W9Pca3w8n+3iof5mvGGOeM8Z8fHhQSYf6p3SkdG6/n8aYzxhjXpT0MUkPybPj75j+ZXw4/v5Q0uestc8f2Jb7scfJztHus9beqfRiYIGkP11we/LwJ0rnIX3oy3EO9+82a+3dSqfOvl/S7y2qYadhrf2gtfY2Sb+rdF7fKyP65/zxZ4y5V2n7Pz3v9zppkL8k6fXGmJIkDX++brh93GPOyKZbrLU9pR/Ejw4felEH/utujLkkKbHWXj/zRs5geFL3zZLeZ62NNb4fzvXxmP4d/Azbkj6jEZ+h0lHQuf9+Wms/K+nHJf2PPDz+sv4ZY27x5Pj7MUl3SHp+ePL2DZK+JOlNyvnYO1GQD8+gPifp/cNN75f0rLX22rjHTvJei2CMuWCMiYZ/DyT9otI+SekKh6ox5p3D339D0t+cfSunZ4z5mKQflvQzwwNDGt8Pp/p4XP+MMRvGmOrw72VJ79X+Z/i4pHuMMW8e/n4u+2eMWTPGvPHA7w9Iui7Ji+NvTP+6Phx/1tpHrbWvs9bebq29Xek/wD+h9H8duR57Ey9ja4z5Y0k/J+m7Jb0q6TVr7VuNMXcoXeK0IamhdImTHb5m5GPnzXH9k/SApL9Veq3gkqR/k/SwtfaV4WveoXQlQEX7y4P+78wbPwVjzFslfVvSv0vaHm5+3lr7s+P64UofR/VP0ieUtj+RtCTpHyU9Yq3dHL7up4fPKUl6VtIHrLU3zrb14xljvkvS30m6oPQ6/9clfcRa+88+HH+j+iepKU+Ov4OGo/KfGi4/zPXY43rkAOA4TnYCgOMIcgBwHEEOAI4jyAHAcQQ5ADiOIAcAxxHkAOA4ghwAHPf/T3Ut1qKPm90AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.violinplot(\n",
    "    [len(seqrec) for seqrec, tax in zip(silva_seqs, silva_taxa)]\n",
    ").set(xlim=(100, 400))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyse taxonomy label distributions at all rank levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter taxonomy. We've analysed the distributions of taxa labels at all levels and have found out that labels starting with lower-case words and labels containig the substring \"sp.\" almost exclusively correspond to unclassified isolates. Remove remaining eukaryotes (theoretically, they should've been discarded at the V4-extraction phase due to primer mismatch)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Picked 1865548 sequences out of 1866614'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def filter_taxonomy(rule: Callable[[str], bool], taxonomy: List[str]) -> List[str]:\n",
    "    \"\"\"\n",
    "    Truncate taxonomies at first occurence of \"bad\" labels\n",
    "    :param rule: a label validation rule; good labels should evaluate as True\n",
    "    \"\"\"\n",
    "    ranks = (rank if rule(rank) else None for rank in taxonomy)\n",
    "    return list(takewhile(bool, ranks))\n",
    "\n",
    "\n",
    "good_label_rule = lambda rank: not (rank.split()[0].islower() or 'sp.' in rank)\n",
    "\n",
    "silva_seqs_noneuk, silva_taxa_noneuk_transformed = zip(*[\n",
    "    (seqrec, filter_taxonomy(good_label_rule, tax)) \n",
    "    for seqrec, tax in zip(silva_seqs, silva_taxa) if not tax[0].startswith('Euk')\n",
    "])\n",
    "f'Picked {len(silva_seqs_noneuk)} sequences out of {len(silva_seqs)}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove sequences lacking genus-level annotation and replace missing species annotations with f'unclassified {genus} isolate'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "silva_seqs_withgen, silva_taxa_withgen = zip(*[\n",
    "    (seqrec, (*tax, f'unclassified {tax[-1]} isolate')[:7]) \n",
    "    for seqrec, tax in zip(silva_seqs_noneuk, silva_taxa_noneuk_transformed) if len(tax) >= 6\n",
    "])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sanity check the most common species labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('unclassified Staphylococcus isolate', 116764),\n",
       " ('unclassified Corynebacterium 1 isolate', 61262),\n",
       " ('unclassified Cutibacterium isolate', 58012),\n",
       " ('unclassified Streptococcus isolate', 55916),\n",
       " ('unclassified Bacteroides isolate', 36719),\n",
       " ('unclassified Pseudomonas isolate', 32403),\n",
       " ('unclassified Bacillus isolate', 24142),\n",
       " ('unclassified WCHB1-32 isolate', 15526),\n",
       " ('unclassified Escherichia-Shigella isolate', 15243),\n",
       " ('unclassified Lactobacillus isolate', 13841),\n",
       " ('unclassified Acinetobacter isolate', 13690),\n",
       " ('unclassified Streptomyces isolate', 11834),\n",
       " ('unclassified Faecalibacterium isolate', 11016),\n",
       " ('unclassified Salmonella isolate', 10010),\n",
       " ('unclassified Lachnospiraceae NK4A136 group isolate', 9426),\n",
       " ('unclassified Blautia isolate', 9370),\n",
       " ('unclassified Burkholderia-Caballeronia-Paraburkholderia isolate', 8678),\n",
       " ('unclassified Cloacibacterium isolate', 8154),\n",
       " ('unclassified Neisseria isolate', 8072),\n",
       " ('unclassified Haemophilus isolate', 7969),\n",
       " ('unclassified Fusobacterium isolate', 7934),\n",
       " ('unclassified Serratia isolate', 7890),\n",
       " ('Streptococcus pneumoniae', 7719),\n",
       " ('unclassified Aquabacterium isolate', 7514),\n",
       " ('unclassified Diaphorobacter isolate', 7412),\n",
       " ('unclassified Clostridium sensu stricto 1 isolate', 7355),\n",
       " ('unclassified Enterobacter isolate', 7260),\n",
       " ('unclassified Veillonella isolate', 7085),\n",
       " ('unclassified Moraxella isolate', 6766),\n",
       " ('Escherichia coli', 6473),\n",
       " ('unclassified Candidatus Nitrosopumilus isolate', 6336),\n",
       " ('unclassified Rothia isolate', 6316),\n",
       " ('unclassified Roseburia isolate', 6315),\n",
       " ('unclassified Gemella isolate', 6127),\n",
       " ('unclassified Prevotella 9 isolate', 5918),\n",
       " ('unclassified Klebsiella isolate', 5855),\n",
       " ('unclassified Lawsonella isolate', 5829),\n",
       " ('unclassified Clade Ia isolate', 5759),\n",
       " ('unclassified Sphingomonas isolate', 5681),\n",
       " ('unclassified Mycobacterium isolate', 5391),\n",
       " ('unclassified Acidovorax isolate', 5384),\n",
       " ('unclassified Enterococcus isolate', 5066),\n",
       " ('unclassified Actinomyces isolate', 5012),\n",
       " ('unclassified Lachnoclostridium isolate', 4935),\n",
       " ('unclassified Anoxybacillus isolate', 4833),\n",
       " ('unclassified Anaerococcus isolate', 4722),\n",
       " ('unclassified Flavobacterium isolate', 4632),\n",
       " ('Staphylococcus aureus', 4600),\n",
       " ('unclassified Vibrio isolate', 4586),\n",
       " ('unclassified Porphyromonas isolate', 4467)]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species = (F(map, op.itemgetter(6)) >> Counter >> Counter.most_common)(silva_taxa_withgen)\n",
    "species[:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export results grouped by taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "51363it [00:18, 2743.42it/s]\n"
     ]
    }
   ],
   "source": [
    "trainset = [\n",
    "    SeqRecord(id=';'.join(['Root', *tax]), name='', description='', seq=seqrec.seq)\n",
    "    for seqrec, tax in zip(silva_seqs_withgen, silva_taxa_withgen)\n",
    "]\n",
    "taxgroups = groupby(sorted(trainset, key=lambda x: x.id), lambda x: x.id)\n",
    "\n",
    "! mkdir -p taxonomy/groups\n",
    "\n",
    "for i, (_, group) in tqdm.tqdm(enumerate(taxgroups, 1)):\n",
    "    with open(f'taxonomy/groups/group{i}.fna', 'w') as groupout:\n",
    "        SeqIO.write(group, groupout, 'fasta')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
